tests/components/melcloud/test_config_flow.py

```diff
@@ -0,0 +1,204 @@
+        )
+        yield mock
+
+
+@pytest.fixture
+def mock_request_info():
+    """Mock RequestInfo to create ClientResposenErrors."""
+    with async_patch("aiohttp.RequestInfo") as mock_ri:
+        mock_ri.return_value.real_url.return_value = ""
+        yield mock_ri
+
+
+async def test_form(hass, mock_login, mock_get_devices):
+    """Test we get the form."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == "form"
+    assert result["errors"] is None
+
+    with async_patch(
+        "homeassistant.components.melcloud.async_setup", return_value=mock_coro(True)
+    ) as mock_setup, async_patch(
+        "homeassistant.components.melcloud.async_setup_entry",
+        return_value=mock_coro(True),
+    ) as mock_setup_entry:
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {"email": "test-email@test-domain.com", "password": "test-password"},
+        )
+
+    assert result2["type"] == "create_entry"
+    assert result2["title"] == "test-email@test-domain.com"
+    assert result2["data"] == {
+        "email": "test-email@test-domain.com",
+        "token": "test-token",
+    }
+    await hass.async_block_till_done()
+    assert len(mock_setup.mock_calls) == 1
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+@pytest.mark.parametrize(
+    "error,reason",
+    [(ClientError(), "cannot_connect"), (asyncio.TimeoutError(), "cannot_connect")],
+)
+async def test_form_errors(hass, mock_login, mock_get_devices, error, reason):
+    """Test we handle cannot connect error."""
+    mock_login.return_value = mock_coro(exception=error)
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with async_patch(
+        "homeassistant.components.melcloud.async_setup", return_value=mock_coro(True)
+    ), async_patch(
+        "homeassistant.components.melcloud.async_setup_entry",
+        return_value=mock_coro(True),
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {"email": "test-email@test-domain.com", "password": "test-password"},
+        )
+
+    assert len(mock_login.mock_calls) == 1
+    assert result2["type"] == "abort"
+    assert result2["reason"] == reason
+
+
+@pytest.mark.parametrize(
+    "error,message",
+    [(401, "invalid_auth"), (403, "invalid_auth"), (500, "cannot_connect")],
+)
+async def test_form_response_errors(
+    hass, mock_login, mock_get_devices, mock_request_info, error, message
+):
+    """Test we handle response errors."""
+    mock_login.return_value = mock_coro(
+        exception=ClientResponseError(mock_request_info(), (), status=error),
+    )
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```

Comment:
We can pass the user input as the `data` keyword argument if we don't need to test the initial form result again.

Reply:
Of course. I wish I had paid enough attention before writing all the tests with the form phase. Fixed.